# LEGO Project

![LegoSchema](Lego_Schema.PNG)

 ## Data

The data has been downloaded directly from their website and includes files for the tables shown in the schema above.  Your first task will be to go through the data and review the schema to understand how the tables all fit together.  There is no data dictionary that I am aware of, but you should be able to understand this data by reviewing the files and schema and visiting their [Download](https://rebrickable.com/downloads/) page and [API](https://rebrickable.com/api/) documentation.  **Note that you must use the files downloaded from Brightspace for this assignment and not download new files from the Rebrickable website.**


In [1]:
import pandas as pd
import numpy as np


pd.set_option('display.max_columns', None)

In [2]:
# import Lego data sets here
# make sure that files are in the same folder as your notebook
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

## Exploration

Take time to look at all of the data sets using `.head()`, `.info()`, `.describe()`, etc to familiarize yourself with the data sets.  As you look at them, notice how the various tables connect together.

In [3]:
# Inventories
print(inventories.head())
print(inventories.info())
print(inventories.describe())

   id  version set_num
0   1        1  7922-1
1   3        1  3931-1
2   4        1  6942-1
3  15        1  5158-1
4  16        1   903-1
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31587 entries, 0 to 31586
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   id       31587 non-null  int64 
 1   version  31587 non-null  int64 
 2   set_num  31587 non-null  object
dtypes: int64(2), object(1)
memory usage: 740.4+ KB
None
                  id       version
count   31587.000000  31587.000000
mean    43538.531389      1.077405
std     32458.542329      0.515993
min         1.000000      1.000000
25%     12193.500000      1.000000
50%     51076.000000      1.000000
75%     62254.500000      1.000000
max    122472.000000     11.000000


In [4]:
# inventory_sets
print(inventory_sets.head())
print(inventory_sets.info())
print(inventory_sets.describe())

   inventory_id  set_num  quantity
0            35  75911-1         1
1            35  75912-1         1
2            39  75048-1         1
3            39  75053-1         1
4            50   4515-1         1
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3413 entries, 0 to 3412
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   inventory_id  3413 non-null   int64 
 1   set_num       3413 non-null   object
 2   quantity      3413 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 80.1+ KB
None
        inventory_id     quantity
count    3413.000000  3413.000000
mean    25839.150601     1.857017
std     31384.971888     6.123060
min        35.000000     1.000000
25%      6117.000000     1.000000
50%     12465.000000     1.000000
75%     27003.000000     1.000000
max    122081.000000    60.000000


In [5]:
#sets
print(sets.head())
print(sets.info())
print(sets.describe())

  set_num                        name  year  theme_id  num_parts
0   001-1                       Gears  1965         1         43
1  0011-2           Town Mini-Figures  1979        67         12
2  0011-3  Castle 2 for 1 Bonus Offer  1987       199          0
3  0012-1          Space Mini-Figures  1979       143         12
4  0013-1          Space Mini-Figures  1979       143         12
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18576 entries, 0 to 18575
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   set_num    18576 non-null  object
 1   name       18576 non-null  object
 2   year       18576 non-null  int64 
 3   theme_id   18576 non-null  int64 
 4   num_parts  18576 non-null  int64 
dtypes: int64(3), object(2)
memory usage: 725.8+ KB
None
               year      theme_id     num_parts
count  18576.000000  18576.000000  18576.000000
mean    2006.200581    405.924365    160.666344
std       13.749798    197.8

In [6]:
#themes
print(themes.head())
print(themes.info())
print(themes.describe())

   id            name  parent_id
0   1         Technic        NaN
1   3     Competition        1.0
2   4  Expert Builder        1.0
3  16      RoboRiders        1.0
4  17  Speed Slammers        1.0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 444 entries, 0 to 443
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   id         444 non-null    int64  
 1   name       444 non-null    object 
 2   parent_id  306 non-null    float64
dtypes: float64(1), int64(1), object(1)
memory usage: 10.5+ KB
None
               id   parent_id
count  444.000000  306.000000
mean   410.659910  346.516340
std    208.165803  198.728717
min      1.000000    1.000000
25%    229.750000  186.000000
50%    438.500000  409.500000
75%    597.250000  507.000000
max    719.000000  697.000000


In [7]:
#inventory_minifigs
print(inventory_minifigs.head())
print(inventory_minifigs.info())
print(inventory_minifigs.describe())

   inventory_id     fig_num  quantity
0             3  fig-001549         1
1             4  fig-000764         1
2            19  fig-000555         1
3            25  fig-000574         1
4            26  fig-000842         1
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18144 entries, 0 to 18143
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   inventory_id  18144 non-null  int64 
 1   fig_num       18144 non-null  object
 2   quantity      18144 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 425.4+ KB
None
        inventory_id      quantity
count   18144.000000  18144.000000
mean    26470.198302      1.066138
std     30578.235290      0.831258
min         3.000000      1.000000
25%      6901.000000      1.000000
50%     13832.000000      1.000000
75%     32633.000000      1.000000
max    122344.000000    100.000000


In [8]:
#minifigs
print(minifigs.head())
print(minifigs.info())
print(minifigs.describe())

      fig_num                             name  num_parts
0  fig-000001               Toy Store Employee          4
1  fig-000002                     Customer Kid          4
2  fig-000003            Assassin Droid, White          8
3  fig-000004                     Basic Figure          4
4  fig-000005  Captain America with Short Legs          3
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11827 entries, 0 to 11826
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   fig_num    11827 non-null  object
 1   name       11827 non-null  object
 2   num_parts  11827 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 277.3+ KB
None
          num_parts
count  11827.000000
mean       4.847637
std        3.562961
min        0.000000
25%        4.000000
50%        4.000000
75%        5.000000
max      148.000000


In [9]:
#inventory_parts
print(inventory_parts.head())
print(inventory_parts.info())
print(inventory_parts.describe())

   inventory_id        part_num  color_id  quantity is_spare
0             1        48379c01        72         1        f
1             1           48395         7         1        f
2             1  stickerupn0077      9999         1        f
3             1         upn0342         0         1        f
4             1         upn0350        25         1        f
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 989529 entries, 0 to 989528
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   inventory_id  989529 non-null  int64 
 1   part_num      989529 non-null  object
 2   color_id      989529 non-null  int64 
 3   quantity      989529 non-null  int64 
 4   is_spare      989529 non-null  object
dtypes: int64(3), object(2)
memory usage: 37.7+ MB
None
        inventory_id       color_id       quantity
count  989529.000000  989529.000000  989529.000000
mean    30715.103387     124.593946       3.308733
std     32809.

In [10]:
#part_categories
print(part_categories.head())
print(part_categories.info())
print(part_categories.describe())

   id                     name
0   1               Baseplates
1   3            Bricks Sloped
2   4  Duplo, Quatro and Primo
3   5           Bricks Special
4   6            Bricks Wedged
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66 entries, 0 to 65
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      66 non-null     int64 
 1   name    66 non-null     object
dtypes: int64(1), object(1)
memory usage: 1.2+ KB
None
              id
count  66.000000
mean   35.363636
std    19.411049
min     1.000000
25%    19.250000
50%    35.500000
75%    51.750000
max    68.000000


In [11]:
#parts
print(parts.head())
print(parts.info())
print(parts.describe())

  part_num                                        name  part_cat_id  \
0   003381                 Sticker Sheet for Set 663-1           58   
1   003383         Sticker Sheet for Sets 618-1, 628-2           58   
2   003402  Sticker Sheet for Sets 310-3, 311-1, 312-3           58   
3   003429                Sticker Sheet for Set 1550-1           58   
4   003432  Sticker Sheet for Sets 357-1, 355-1, 940-1           58   

  part_material  
0       Plastic  
1       Plastic  
2       Plastic  
3       Plastic  
4       Plastic  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44781 entries, 0 to 44780
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   part_num       44781 non-null  object
 1   name           44781 non-null  object
 2   part_cat_id    44781 non-null  int64 
 3   part_material  44781 non-null  object
dtypes: int64(1), object(3)
memory usage: 1.4+ MB
None
        part_cat_id
count  44781.000000
mean

In [12]:
#colors
print(colors.head())
print(colors.info())
print(colors.describe())

   id            name     rgb is_trans
0  -1       [Unknown]  0033B2        f
1   0           Black  05131D        f
2   1            Blue  0055BF        f
3   2           Green  237841        f
4   3  Dark Turquoise  008F9B        f
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190 entries, 0 to 189
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        190 non-null    int64 
 1   name      190 non-null    object
 2   rgb       190 non-null    object
 3   is_trans  190 non-null    object
dtypes: int64(1), object(3)
memory usage: 6.1+ KB
None
                id
count   190.000000
mean    479.394737
std     820.735479
min      -1.000000
25%      61.250000
50%     182.500000
75%    1015.750000
max    9999.000000


In [13]:
#part_relationships
print(part_relationships.head())
print(part_relationships.info())
print(part_relationships.describe())

  rel_type child_part_num parent_part_num
0        P    87079pr9974           87079
1        P     3960pr9971            3960
2        R    98653pr0003     98086pr0003
3        R    98653pr0003    98088pat0003
4        R    98653pr0003    98089pat0003
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24508 entries, 0 to 24507
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   rel_type         24508 non-null  object
 1   child_part_num   24508 non-null  object
 2   parent_part_num  24508 non-null  object
dtypes: object(3)
memory usage: 574.5+ KB
None
       rel_type child_part_num parent_part_num
count     24508          24508           24508
unique        6          22339            3858
top           P           3814           3626c
freq      19745             23            2353


In [14]:
#elements
print(elements.head())
print(elements.info())
print(elements.describe())

  element_id  part_num  color_id
0    6300211  67906c01        14
1    4566309      2564         0
2    4275423     53657      1004
3    6194308     92926        71
4    6229123     26561         4
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62684 entries, 0 to 62683
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   element_id  62684 non-null  object
 1   part_num    62684 non-null  object
 2   color_id    62684 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 1.4+ MB
None
           color_id
count  62684.000000
mean     608.309983
std     2210.989312
min       -1.000000
25%       14.000000
50%       30.000000
75%      148.000000
max     9999.000000


## Assignment

Answer all the questions below using the available data.  There are multiple ways to get some of the answers.  I would suggest that you attempt to merge some data sets as needed to assist with getting answers.  This will make your tasks a lot easier and your code will be cleaner.

Also, you will be creating functions to get the answers to some of these questions.  This will make it much easier for you to get answers to similar questions in the future and CodeGrade will be checking your functions using different inputs.  

**Q1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer).  Nothing else should be returned.

In [15]:
def data_count(df):
    return len(df)

**Code Check:**  If you call your `data_count` function using the `sets` DataFrame, you should see that there are 18,576 sets represented in the data.

In [16]:
data_count(sets)

18576

**Q2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Save this as `Q2`.

In [17]:
Q2 = round(sets.num_parts.mean())
Q2

161

**Q3:** What is the median number of parts per set (output as an integer)?  Save this as `Q3`.

In [18]:
Q3 = round(sets.num_parts.median())
Q3

37

You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Q4:** What is the largest number of parts that a set has? Save this as `Q4`.

In [19]:
Q4 = sets.num_parts.max()
Q4

11695

**Q5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `Q5`.  Do not reset the index.

In [20]:
Q5 = sets.loc[(sets['num_parts'] == 11695)]
Q5

set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695

**Q6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [21]:
def select_set_row(val):
    return sets.loc[sets['set_num']== val]
    

In [22]:
#select_set_row('001-1')


**Q7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [23]:
def select_set_numparts(num_parts):
    return sets[sets['num_parts'] == num_parts].sort_values(['year','set_num'])



**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

In [24]:
select_set_numparts(0)

set_num                                               name  \
11061          700.1-2                     Extra-Large Gift Set (Mursten)   
17563         b55dk-01               Lego Mursten - System i Leg Byggebog   
17564         b56de-01                                    System im Spiel   
11072          700.2-2                                       Gift Package   
11078         700.3A-2                                       Gift Package   
...                ...                                                ...   
15619         854195-1                                    Ernie Key Chain   
15620         854197-1                               Demogorgon Key Chain   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   

       year  theme_id  num_parts  
11061  1953       366          0  
17563  1955       497          0  
17564  1956       497          0  
11072  1957       366          0  
11078  1957       372          0  
...     ...       ...        ...  
15619  2022       503          0  
15620  2022       503          0  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  

[3146 rows x 5 columns]

**Q8:** 
- The oldest set listed was made in what year?  Save the year as `Q8A` (as an integer).  
- The newest set listed was made in what year? Save the year as `Q8B` (as an integer).

In [25]:
Q8A = sets['year'].min()
Q8A

1949

In [26]:
Q8B = sets['year'].max()
Q8B

2022

**Q9:** 
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.
- Use the above function to return a DataFrame for rows that match the minimum year (calculated as `Q8A` above) and call this `Q9A`.  
- Then use the above function to return a DataFrame for rows that match the maximum year (calculated as `Q8B` above) and call this `Q9B`.

In [27]:
def select_set_year(year):
    return sets[sets['year'] == year].sort_values(['set_num'])

In [28]:
Q9A= select_set_year(1949)

Q9B = select_set_year(2022)


**Q10:** What is the median year represented in the `sets` data?  Save this as `Q10` (as an integer).

In [29]:
Q10 = int(sets.year.median())
Q10

2010

Notice that the first set was back in 1949 but over half of the sets were after the median year. This could be an indication that the number of sets has increased greatly over the past number of years.

**Q11:** What is the average number of sets listed per year?  Save this as `Q11` (as an integer).

In [30]:
Q11 = int(sets.groupby(['year'])[['set_num']].count().mean())
Q11

258

**Q12:** 
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [31]:
themes_2 = themes.copy()
themes_2.rename(columns={'name':'name_themes'}, inplace = True)
themes_2

id                 name_themes  parent_id
0      1                     Technic        NaN
1      3                 Competition        1.0
2      4              Expert Builder        1.0
3     16                  RoboRiders        1.0
4     17              Speed Slammers        1.0
..   ...                         ...        ...
439  715                      Marvel      535.0
440  716                     Modulex        NaN
441  717                 Speed Racer        NaN
442  718       Series 22 Minifigures      535.0
443  719  BrickLink Designer Program        NaN

[444 rows x 3 columns]

In [32]:
sets_2 = sets.copy()
sets_2.rename(columns={'theme_id':'id'}, inplace = True)
sets_2

set_num                             name  year   id  num_parts
0            001-1                            Gears  1965    1         43
1           0011-2                Town Mini-Figures  1979   67         12
2           0011-3       Castle 2 for 1 Bonus Offer  1987  199          0
3           0012-1               Space Mini-Figures  1979  143         12
4           0013-1               Space Mini-Figures  1979  143         12
...            ...                              ...   ...  ...        ...
18571      XWING-1              Mini X-Wing Fighter  2019  158         60
18572      XWING-2                X-Wing Trench Run  2019  158         52
18573  YODACHRON-1  Yoda Chronicles Promotional Set  2013  158        413
18574   YTERRIER-1                Yorkshire Terrier  2018  598          0
18575     ZX8000-1             ZX 8000 LEGO Sneaker  2020  501          0

[18576 rows x 5 columns]

In [33]:
sets_1 = pd.merge(sets_2, themes_2, on = 'id', how ='left')
sets_1

set_num                             name  year   id  num_parts  \
0            001-1                            Gears  1965    1         43   
1           0011-2                Town Mini-Figures  1979   67         12   
2           0011-3       Castle 2 for 1 Bonus Offer  1987  199          0   
3           0012-1               Space Mini-Figures  1979  143         12   
4           0013-1               Space Mini-Figures  1979  143         12   
...            ...                              ...   ...  ...        ...   
18571      XWING-1              Mini X-Wing Fighter  2019  158         60   
18572      XWING-2                X-Wing Trench Run  2019  158         52   
18573  YODACHRON-1  Yoda Chronicles Promotional Set  2013  158        413   
18574   YTERRIER-1                Yorkshire Terrier  2018  598          0   
18575     ZX8000-1             ZX 8000 LEGO Sneaker  2020  501          0   

        name_themes  parent_id  
0           Technic        NaN  
1      Classic Town       50.0  
2      Lion Knights      186.0  
3      Supplemental      126.0  
4      Supplemental      126.0  
...             ...        ...  
18571     Star Wars        NaN  
18572     Star Wars        NaN  
18573     Star Wars        NaN  
18574   Promotional        NaN  
18575          Gear        NaN  

[18576 rows x 7 columns]

In [34]:
def theme_by_year(val):
    return sets_1[sets_1['year'] == val].sort_values(['id']).reset_index()[['id','name_themes']].drop_duplicates(keep='first').reset_index(drop = True)

In [35]:
theme_by_year(1960)

id   name_themes
0  371  Supplemental
1  497         Books
2  513       Classic

**Code Check:** Call the `theme_by_year()` function on the year `1960`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

In [36]:
theme_by_year(1997)

id             name_themes
0     1                 Technic
1    67            Classic Town
2    86                  Divers
3    89                 Outback
4    90                Paradisa
5    94                Town Jr.
6   138               RoboForce
7   144                     UFO
8   148               Pirates I
9   195          Fright Knights
10  227               Christmas
11  236                      9V
12  278              Model Team
13  309           Aquaraiders I
14  318                Belville
15  399               Freestyle
16  408        LEGO Brand Store
17  443           Service Packs
18  453                 Technic
19  464           Time Cruisers
20  467                   Basic
21  468             Basic Model
22  469               Basic Set
23  473            Supplemental
24  475                 Western
25  497                   Books
26  501                    Gear
27  503               Key Chain
28  505               Basic Set
29  507   Educational and Dacta
30  512                  Castle
31  516       Duplo and Explore
32  528            Supplemental
33  529                 Technic
34  530             Control Lab
35  532            Supplemental
36  534  Universal Building Set
37  632                    Town
38  652               Playhouse
39  658                    Dino
40  668                   Scala

**Q13:** What theme has the most number of sets associated with it?  Save the theme name as `Q13` (as a string).

In [37]:
sets_1.groupby(['name_themes'])[['name_themes']].size().nlargest(1)
Q13 = 'Gear'
Q13

'Gear'

**Q14:** 
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

In [38]:
sets_3 = sets_1.copy()
sets_3.rename(columns={'name':'name_sets'}, inplace = True)
sets_3.rename(columns={'id':'theme_id'}, inplace = True)
sets_3

set_num                        name_sets  year  theme_id  \
0            001-1                            Gears  1965         1   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-3       Castle 2 for 1 Bonus Offer  1987       199   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18571      XWING-1              Mini X-Wing Fighter  2019       158   
18572      XWING-2                X-Wing Trench Run  2019       158   
18573  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18574   YTERRIER-1                Yorkshire Terrier  2018       598   
18575     ZX8000-1             ZX 8000 LEGO Sneaker  2020       501   

       num_parts   name_themes  parent_id  
0             43       Technic        NaN  
1             12  Classic Town       50.0  
2              0  Lion Knights      186.0  
3             12  Supplemental      126.0  
4             12  Supplemental      126.0  
...          ...           ...        ...  
18571         60     Star Wars        NaN  
18572         52     Star Wars        NaN  
18573        413     Star Wars        NaN  
18574          0   Promotional        NaN  
18575          0          Gear        NaN  

[18576 rows x 7 columns]

In [39]:
def theme_by_name(val):
    return sets_3[sets_3['name_themes']== val].sort_values(['set_num']).reset_index()[['set_num','name_sets','year','num_parts', 'theme_id','name_themes']]

**Code Check:** Call the `theme_by_name()` function using the theme name from `Q13`.  You should return 1,904 sets that match that theme name.

In [40]:
theme_by_name('Gear')

set_num                                          name_sets  \
0              11138-1                            Bionicle Barraki Wallet   
1       119645111436-1                Classic Space Astronaut Plush (Red)   
2       119645111504-1               Classic Space Astronaut Plush (Blue)   
3       119645111740-1              Classic Space Astronaut Plush (White)   
4       119645111818-1               Classic Space Astronaut Plush (Pink)   
...                ...                                                ...   
1899           WATCH-3                   Watch System Voyager (UFO Space)   
1900         WIIGAME-1                              Bionicle Heroes - Wii   
1901         WIIGAME-2  Indiana Jones: The Original Adventures Video G...   
1902  WOODENELEPHANT-1  Original Wooden Elephant Pull-Along Toy and Print   
1903          ZX8000-1                               ZX 8000 LEGO Sneaker   

      year  num_parts  theme_id name_themes  
0     2008          0       501        Gear  
1     2021          0       501        Gear  
2     2021          0       501        Gear  
3     2021          0       501        Gear  
4     2021          0       501        Gear  
...    ...        ...       ...         ...  
1899  1997         29       501        Gear  
1900  2007          0       501        Gear  
1901  2008          0       501        Gear  
1902  2021          0       501        Gear  
1903  2020          0       501        Gear  

[1904 rows x 6 columns]

**Q15:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [41]:
def theme_by_setnum(val):
    return sets_3[sets_3['set_num']== val]['name_themes'].max()

**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [42]:
theme_by_setnum('11138-1')

'Gear'

**Q16:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You **do not** need to save this as `Q16`.  Instead you will use this information for later questions.

In [43]:
MF= sets_3[sets_3['name_sets']=='Millennium Falcon'].nlargest(1, columns ='num_parts')
MF

set_num          name_sets  year  theme_id  num_parts  \
12884  75192-1  Millennium Falcon  2017       171       7541   

                     name_themes  parent_id  
12884  Ultimate Collector Series      158.0

**Q17:** 
- How many minifigs are contained in the set from `Q16`?  Save this as `Q17` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [44]:
sets_4 = sets_3.copy()
inventories_1 = inventories.copy()
inventories_1.rename(columns={'id':'inventory_id'}, inplace = True)

sets_5 = pd.merge(sets_4, inventories_1)
sets_5

set_num                        name_sets  year  theme_id  \
0            001-1                            Gears  1965         1   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-3       Castle 2 for 1 Bonus Offer  1987       199   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
19755      XWING-1              Mini X-Wing Fighter  2019       158   
19756      XWING-2                X-Wing Trench Run  2019       158   
19757  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
19758   YTERRIER-1                Yorkshire Terrier  2018       598   
19759     ZX8000-1             ZX 8000 LEGO Sneaker  2020       501   

       num_parts   name_themes  parent_id  inventory_id  version  
0             43       Technic        NaN         24696        1  
1             12  Classic Town       50.0          5087        1  
2              0  Lion Knights      186.0          2216        1  
3             12  Supplemental      126.0          1414        1  
4             12  Supplemental      126.0          4609        1  
...          ...           ...        ...           ...      ...  
19755         60     Star Wars        NaN         35852        1  
19756         52     Star Wars        NaN         37243        1  
19757        413     Star Wars        NaN         36159        1  
19758          0   Promotional        NaN         88108        1  
19759          0          Gear        NaN         74972        1  

[19760 rows x 9 columns]

In [45]:
inventory_minifigs_1 = inventory_minifigs.copy()
sets_6 = pd.merge(sets_5, inventory_minifigs_1)
sets_6

set_num                        name_sets  year  theme_id  \
0           0011-2                Town Mini-Figures  1979        67   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-2                Town Mini-Figures  1979        67   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18139  WEETABIX4-1                    House and Car  1976       413   
18140    WILLIAM-1                        Will.i.am  2016       535   
18141  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18142  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18143  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   

       num_parts              name_themes  parent_id  inventory_id  version  \
0             12             Classic Town       50.0          5087        1   
1             12             Classic Town       50.0          5087        1   
2             12             Classic Town       50.0          5087        1   
3             12             Supplemental      126.0          1414        1   
4             12             Supplemental      126.0          4609        1   
...          ...                      ...        ...           ...      ...   
18139        147                 Building      411.0         13813        1   
18140          6  Collectible Minifigures        NaN         34919        1   
18141        413                Star Wars        NaN         36159        1   
18142        413                Star Wars        NaN         36159        1   
18143        413                Star Wars        NaN         36159        1   

          fig_num  quantity  
0      fig-000225         1  
1      fig-000834         1  
2      fig-004401         1  
3      fig-000020         2  
4      fig-001127         2  
...           ...       ...  
18139  fig-001361         1  
18140  fig-010738         1  
18141  fig-004483         1  
18142  fig-004484         1  
18143  fig-010456         1  

[18144 rows x 11 columns]

In [46]:
minifigs_1 = minifigs.copy()
minifigs_1.rename(columns={'name':'fig_name','num_parts':'fig_num_parts'}, inplace = True)

sets_7 = pd.merge(sets_6, minifigs_1, how='left')
sets_7

set_num                        name_sets  year  theme_id  \
0           0011-2                Town Mini-Figures  1979        67   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-2                Town Mini-Figures  1979        67   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18139  WEETABIX4-1                    House and Car  1976       413   
18140    WILLIAM-1                        Will.i.am  2016       535   
18141  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18142  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18143  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   

       num_parts              name_themes  parent_id  inventory_id  version  \
0             12             Classic Town       50.0          5087        1   
1             12             Classic Town       50.0          5087        1   
2             12             Classic Town       50.0          5087        1   
3             12             Supplemental      126.0          1414        1   
4             12             Supplemental      126.0          4609        1   
...          ...                      ...        ...           ...      ...   
18139        147                 Building      411.0         13813        1   
18140          6  Collectible Minifigures        NaN         34919        1   
18141        413                Star Wars        NaN         36159        1   
18142        413                Star Wars        NaN         36159        1   
18143        413                Star Wars        NaN         36159        1   

          fig_num  quantity  \
0      fig-000225         1   
1      fig-000834         1   
2      fig-004401         1   
3      fig-000020         2   
4      fig-001127         2   
...           ...       ...   
18139  fig-001361         1   
18140  fig-010738         1   
18141  fig-004483         1   
18142  fig-004484         1   
18143  fig-010456         1   

                                                fig_name  fig_num_parts  
0      Policeman, Black Suit with Pocket and Badge, W...              4  
1                Man, Red Torso, Blue Legs, Red Hard Hat              4  
2                           Woman - Blue Torso, Red Legs              4  
3      Classic Spaceman, Red with Airtanks (3842a Hel...              5  
4      Classic Spaceman, White with Airtanks (3842a H...              5  
...                                                  ...            ...  
18139         LEGOLAND White Torso, Black Legs, Red Hair              4  
18140                                          Will.i.am              4  
18141  Yoda, Sand Green Head with Gray Hair, Dark Ora...              3  
18142                                    Padawan, Female              4  
18143                                     Holocron Droid             15  

[18144 rows x 13 columns]

In [47]:
Q17 = sets_7[sets_7['set_num']=='75192-1'][['quantity']].sum().max()
Q17

8

**Q18:** 
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [48]:
sets_7

set_num                        name_sets  year  theme_id  \
0           0011-2                Town Mini-Figures  1979        67   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-2                Town Mini-Figures  1979        67   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18139  WEETABIX4-1                    House and Car  1976       413   
18140    WILLIAM-1                        Will.i.am  2016       535   
18141  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18142  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18143  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   

       num_parts              name_themes  parent_id  inventory_id  version  \
0             12             Classic Town       50.0          5087        1   
1             12             Classic Town       50.0          5087        1   
2             12             Classic Town       50.0          5087        1   
3             12             Supplemental      126.0          1414        1   
4             12             Supplemental      126.0          4609        1   
...          ...                      ...        ...           ...      ...   
18139        147                 Building      411.0         13813        1   
18140          6  Collectible Minifigures        NaN         34919        1   
18141        413                Star Wars        NaN         36159        1   
18142        413                Star Wars        NaN         36159        1   
18143        413                Star Wars        NaN         36159        1   

          fig_num  quantity  \
0      fig-000225         1   
1      fig-000834         1   
2      fig-004401         1   
3      fig-000020         2   
4      fig-001127         2   
...           ...       ...   
18139  fig-001361         1   
18140  fig-010738         1   
18141  fig-004483         1   
18142  fig-004484         1   
18143  fig-010456         1   

                                                fig_name  fig_num_parts  
0      Policeman, Black Suit with Pocket and Badge, W...              4  
1                Man, Red Torso, Blue Legs, Red Hard Hat              4  
2                           Woman - Blue Torso, Red Legs              4  
3      Classic Spaceman, Red with Airtanks (3842a Hel...              5  
4      Classic Spaceman, White with Airtanks (3842a H...              5  
...                                                  ...            ...  
18139         LEGOLAND White Torso, Black Legs, Red Hair              4  
18140                                          Will.i.am              4  
18141  Yoda, Sand Green Head with Gray Hair, Dark Ora...              3  
18142                                    Padawan, Female              4  
18143                                     Holocron Droid             15  

[18144 rows x 13 columns]

In [49]:
def get_minifigs(set_num, ver):
    set_num = str(set_num)
    ver= int(ver)
    return sets_7[sets_7['set_num']== set_num][sets_7['version']== ver][['set_num','fig_num','fig_name','fig_num_parts']].sort_values(by=(['fig_num'])).reset_index(drop=True)

**Q19:** Using the above function, output information about the minifigs from the set discussed in `Q16`.  Save this output as `Q19`.

In [50]:
Q19 = get_minifigs('75192-1',1)
Q19

<ipython-input-49-50a56c1654d0>:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  return sets_7[sets_7['set_num']== set_num][sets_7['version']== ver][['set_num','fig_num','fig_name','fig_num_parts']].sort_values(by=(['fig_num'])).reset_index(drop=True)


set_num     fig_num                                           fig_name  \
0  75192-1  fig-001714                              Chewbacca, Dark Brown   
1  75192-1  fig-001810                             Finn, Black Undershirt   
2  75192-1  fig-002057                                 Rey, Dark Tan Robe   
3  75192-1  fig-002514    C-3PO, Pearl Gold, Colorful Wires, Printed Legs   
4  75192-1  fig-002544                                               BB-8   
5  75192-1  fig-002549  Princess Leia, Hoth Outfit, White Jacket, Prin...   
6  75192-1  fig-002550  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...   
7  75192-1  fig-002551                               Han Solo, Old, Angry   

   fig_num_parts  
0              3  
1              4  
2              4  
3              3  
4              2  
5              4  
6              4  
7              4

**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [51]:
get_minifigs('10217-1',1)

<ipython-input-49-50a56c1654d0>:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  return sets_7[sets_7['set_num']== set_num][sets_7['version']== ver][['set_num','fig_num','fig_name','fig_num_parts']].sort_values(by=(['fig_num'])).reset_index(drop=True)


set_num     fig_num                                           fig_name  \
0   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
1   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
2   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               6  
1               3  
2               5  
3               4  
4               4  
5               4  
6               4  
7               4  
8               3  
9               3  
10              4

**Q20:** 
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`. 
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [52]:
sets_7

set_num                        name_sets  year  theme_id  \
0           0011-2                Town Mini-Figures  1979        67   
1           0011-2                Town Mini-Figures  1979        67   
2           0011-2                Town Mini-Figures  1979        67   
3           0012-1               Space Mini-Figures  1979       143   
4           0013-1               Space Mini-Figures  1979       143   
...            ...                              ...   ...       ...   
18139  WEETABIX4-1                    House and Car  1976       413   
18140    WILLIAM-1                        Will.i.am  2016       535   
18141  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18142  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   
18143  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158   

       num_parts              name_themes  parent_id  inventory_id  version  \
0             12             Classic Town       50.0          5087        1   
1             12             Classic Town       50.0          5087        1   
2             12             Classic Town       50.0          5087        1   
3             12             Supplemental      126.0          1414        1   
4             12             Supplemental      126.0          4609        1   
...          ...                      ...        ...           ...      ...   
18139        147                 Building      411.0         13813        1   
18140          6  Collectible Minifigures        NaN         34919        1   
18141        413                Star Wars        NaN         36159        1   
18142        413                Star Wars        NaN         36159        1   
18143        413                Star Wars        NaN         36159        1   

          fig_num  quantity  \
0      fig-000225         1   
1      fig-000834         1   
2      fig-004401         1   
3      fig-000020         2   
4      fig-001127         2   
...           ...       ...   
18139  fig-001361         1   
18140  fig-010738         1   
18141  fig-004483         1   
18142  fig-004484         1   
18143  fig-010456         1   

                                                fig_name  fig_num_parts  
0      Policeman, Black Suit with Pocket and Badge, W...              4  
1                Man, Red Torso, Blue Legs, Red Hard Hat              4  
2                           Woman - Blue Torso, Red Legs              4  
3      Classic Spaceman, Red with Airtanks (3842a Hel...              5  
4      Classic Spaceman, White with Airtanks (3842a H...              5  
...                                                  ...            ...  
18139         LEGOLAND White Torso, Black Legs, Red Hair              4  
18140                                          Will.i.am              4  
18141  Yoda, Sand Green Head with Gray Hair, Dark Ora...              3  
18142                                    Padawan, Female              4  
18143                                     Holocron Droid             15  

[18144 rows x 13 columns]

In [53]:
def minifigs_from_themes(inp):
    return sets_7[sets_7['theme_id']== inp].drop_duplicates(['fig_num']).sort_values(['fig_num'])[['fig_num','fig_name']].reset_index(drop = True)

**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in `Q16`.  There should be 149 minifigs output for that respective theme.

In [54]:
minifigs_from_themes(171)

fig_num                                           fig_name
0    fig-000003                              Assassin Droid, White
1    fig-000124                           Chewbacca, Reddish Brown
2    fig-000144  Luke Skywalker, White Robe, White Legs, Cheek ...
3    fig-000303           Clone Trooper, Phase I Armor, Brown Eyes
4    fig-000516  Darth Vader, Light Bluish Gray Skin, Scarred R...
..          ...                                                ...
144  fig-011942  AT-AT Driver - Light Bluish Grey Helmet, Sand ...
145  fig-011944  Snowtrooper, White Hips, Reddish Brown Head, F...
146  fig-011945  Snowtrooper, White Hips, Light Flesh Head, Female
147  fig-011946  Snowtrooper, White Hips, Light Flesh Head, Off...
148  fig-011957    Snowtrooper, White Hips, Light Flesh Head, Grin

[149 rows x 2 columns]

**Q21:** 
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [55]:
def sets_from_minifig(mfn):
    return sets_7[sets_7['fig_num']== mfn].sort_values(['set_num','version'])[['set_num','version','name_sets','fig_num','fig_name']].reset_index(drop = True)

**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [56]:
sets_from_minifig('fig-001039')

set_num  version            name_sets     fig_num  \
0  30100-1        1  Andrea on the Beach  fig-001039   
1  30107-1        1       Birthday Party  fig-001039   
2   3061-1        1       City Park Café  fig-001039   

                                            fig_name  
0  Andrea - Bright Light Orange Top, Light Aqua S...  
1  Andrea - Bright Light Orange Top, Light Aqua S...  
2  Andrea - Bright Light Orange Top, Light Aqua S...

**Q22:** 

*Note:  Before you begin working on this question, please note that CodeGrade has a 1GB RAM limit.  If you see an `Exit -9` code when submitting your assignment, it is probably due to this RAM limit and previous students had an issue with this question causing too much memory usage.  The biggest thing to think about is to only use the tables that you need to answer for this question.  For instance, you do not need any of the minifigs data since this is not needed for the output and the merging of this data could cause the number of rows to "explode" in terms of the total count.  I understand that this can be annoying that we have to work around CodeGrade's limitations, but it is good practice in memory management.*

- Create a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [64]:
parts_2 = parts.copy()
parts_2.rename(columns={'name':'name_part'}, inplace = True)
parts_2

part_num                                   name_part  part_cat_id  \
0           003381                 Sticker Sheet for Set 663-1           58   
1           003383         Sticker Sheet for Sets 618-1, 628-2           58   
2           003402  Sticker Sheet for Sets 310-3, 311-1, 312-3           58   
3           003429                Sticker Sheet for Set 1550-1           58   
4           003432  Sticker Sheet for Sets 357-1, 355-1, 940-1           58   
...            ...                                         ...          ...   
44776  znapupn0002       Znap Connector 3 x 3 - 4 way B (Beam)           43   
44777  znapupn0003              Znap Connector 1 x 3 - 2 way A           43   
44778  znapupn0004                         Znap Beam 3, 1 Hole           43   
44779  znapupn0005     Znap Connector 3 x 3 - 4 way C (Closed)           43   
44780  znapupn0006           Wheel 68mm Znap Propeller (9 x 2)           29   

      part_material  
0           Plastic  
1           Plastic  
2           Plastic  
3           Plastic  
4           Plastic  
...             ...  
44776       Plastic  
44777       Plastic  
44778       Plastic  
44779       Plastic  
44780       Plastic  

[44781 rows x 4 columns]

In [72]:
colors_2 = colors.copy()
colors_2.rename(columns={'name':'name_color','id':'color_id'}, inplace = True)
colors_2

color_id            name_color     rgb is_trans
0          -1             [Unknown]  0033B2        f
1           0                 Black  05131D        f
2           1                  Blue  0055BF        f
3           2                 Green  237841        f
4           3        Dark Turquoise  008F9B        f
..        ...                   ...     ...      ...
185      1059     Trans-Purple Opal  8320B7        t
186      1060      Trans-Green Opal  84B68D        t
187      1061  Trans-Dark Blue Opal  0020A0        t
188      1062        Vibrant Yellow  EBD800        f
189      9999  [No Color/Any Color]  05131D        f

[190 rows x 4 columns]

In [78]:
sets_22 = sets_5.copy()
sets_23 = pd.merge(sets_22, inventory_parts, how='right')
sets_24 = pd.merge(sets_23, parts_2, how='right')
sets_25 = pd.merge(sets_24, colors_2, how ='right')
sets_25= sets_25[['set_num','version','name_sets','year','name_themes','part_num','name_part','name_color','quantity','is_spare']]
sets_25[['version','year','quantity']] = sets_25[['version','year','quantity']].fillna(0).astype(int)
sets_25

set_num  version                          name_sets  year  \
0       LEGO-Modulex-1        4  Unused Modulex parts sold by LEGO  1963   
1       LEGO-Modulex-1        4  Unused Modulex parts sold by LEGO  1963   
2       LEGO-Modulex-1        4  Unused Modulex parts sold by LEGO  1963   
3       LEGO-Modulex-1        4  Unused Modulex parts sold by LEGO  1963   
4               4296-1        1                       Green Strata  2004   
...                ...      ...                                ...   ...   
989535         71766-1        1           Lloyd’s Legendary Dragon  2022   
989536         65580-1        1        Knights' Kingdom Value Pack  2004   
989537            90-2        1            Educational Box - Empty  1965   
989538          700L-1        1        Empty Kindergarten LEGO Box  1961   
989539          700K-1        1              Kindergarten LEGO Set  1960   

               name_themes    part_num  \
0                  Modulex       1012A   
1                  Modulex       1022B   
2                  Modulex       1032A   
3                  Modulex       1051A   
4                Basic Set        3437   
...                    ...         ...   
989535             Ninjago     upn0492   
989536  Knights Kingdom II  vladekmask   
989537             Classic      wood03   
989538             Classic      wood04   
989539             Classic      wood04   

                                                name_part  \
0           Modulex Tile 2 x 12 without Internal Supports   
1               Modulex Tile 2 x 2 with Internal Supports   
2            Modulex Tile 2 x 3 without Internal Supports   
3            Modulex Tile 1 x 5 without Internal Supports   
4                                       Duplo Brick 2 x 2   
...                                                   ...   
989535  Wing, Right, Dragon, Azure Bones, Green/Dark G...   
989536  Play Gear, Mask, Hard Plastic, Knights Kingdom...   
989537           Storage Box, Wooden with Red Sliding Top   
989538  Storage Box, Wooden with Plain Sliding Top and...   
989539  Storage Box, Wooden with Plain Sliding Top and...   

                  name_color  quantity is_spare  
0                  [Unknown]         1        f  
1                  [Unknown]         1        f  
2                  [Unknown]         1        f  
3                  [Unknown]         1        f  
4                  [Unknown]         4        f  
...                      ...       ...      ...  
989535  [No Color/Any Color]         1        f  
989536  [No Color/Any Color]         1        f  
989537  [No Color/Any Color]         1        f  
989538  [No Color/Any Color]         1        f  
989539  [No Color/Any Color]         1        f  

[989540 rows x 10 columns]

In [75]:
def set_parts(set_num, ver):
    set_num = str(set_num)
    ver= int(ver)
    return sets_25[sets_25['set_num']== set_num][sets_25['version']== ver].sort_values(by=(['part_num','name_color'])).reset_index(drop=True)

**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in `Q16`. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [76]:
set_parts('75192-1',1)

<ipython-input-75-cda553b9c747>:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  return sets_25[sets_25['set_num']== set_num][sets_25['version']== ver].sort_values(by=(['part_num','name_color'])).reset_index(drop=True)


set_num  version          name_sets    year                name_themes  \
0    75192-1      1.0  Millennium Falcon  2017.0  Ultimate Collector Series   
1    75192-1      1.0  Millennium Falcon  2017.0  Ultimate Collector Series   
2    75192-1      1.0  Millennium Falcon  2017.0  Ultimate Collector Series   
3    75192-1      1.0  Millennium Falcon  2017.0  Ultimate Collector Series   
4    75192-1      1.0  Millennium Falcon  2017.0  Ultimate Collector Series   
..       ...      ...                ...     ...                        ...   
725  75192-1      1.0  Millennium Falcon  2017.0  Ultimate Collector Series   
726  75192-1      1.0  Millennium Falcon  2017.0  Ultimate Collector Series   
727  75192-1      1.0  Millennium Falcon  2017.0  Ultimate Collector Series   
728  75192-1      1.0  Millennium Falcon  2017.0  Ultimate Collector Series   
729  75192-1      1.0  Millennium Falcon  2017.0  Ultimate Collector Series   

    part_num                                          name_part  \
0      10197  Technic Pin Connector Hub with 2 Perpendicular...   
1      10201            Bracket 1 x 2 - 1 x 4 [Rounded Corners]   
2      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
3      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
4      10314  Brick Curved 1 x 4 x 1 1/3 No Studs, Curved To...   
..       ...                                                ...   
725    99563  Tile Special 1 x 2 with Sloped Walls AKA Money...   
726    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
727    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
728    99781                              Bracket 1 x 2 - 1 x 2   
729    99784  Bar 1 x 12 with 1 x 2 Plate End with Hollow St...   

            name_color  quantity is_spare  
0     Dark Bluish Gray       3.0        f  
1    Light Bluish Gray      12.0        f  
2     Dark Bluish Gray       5.0        f  
3                White      22.0        f  
4                  Tan       3.0        f  
..                 ...       ...      ...  
725  Light Bluish Gray      70.0        f  
726              Black       6.0        f  
727   Dark Bluish Gray      25.0        f  
728  Light Bluish Gray       2.0        f  
729  Light Bluish Gray       3.0        f  

[730 rows x 10 columns]

  **Q23:**  Get a list of all unique color names, sort by color name, and output as a list.  Save this as `Q23`. 

In [ ]:
Q23 = colors.sort_values(by = 'name', ascending = True)['name'].tolist()



Q23

**Q24:** How many total transparent colors are in the `colors` DataFrame?  Save this as `Q24` (as an integer).

In [ ]:
#colors.groupby(['is_trans'])[['name']].count()
Q24 = colors[colors.is_trans == "t"].count()['is_trans']
#Q24 = colors.groupby('is_trans').count()

Q24

**Q25:**  
- Create a function called `get_part_colors` that takes as input a part number (as a string) and returns a DataFrame of all colors a part has appeared in.  
- The DataFrame should be sorted by color id  
- The index should go from 0 to n-1.

In [95]:
inventory_parts_25 = inventory_parts.copy()
inventory_parts_25.rename(columns={'color_id':'id'}, inplace = True) 

colors_25 = colors.copy()

colors_final= pd.merge(inventory_parts_25, colors_25, how='left')
colors_final

inventory_id        part_num    id  quantity is_spare  \
0                  1        48379c01    72         1        f   
1                  1           48395     7         1        f   
2                  1  stickerupn0077  9999         1        f   
3                  1         upn0342     0         1        f   
4                  1         upn0350    25         1        f   
...              ...             ...   ...       ...      ...   
989524        122472           99206    71         2        f   
989525        122472           99207    71         2        f   
989526        122472           99207     0         2        f   
989527        122472           99780     0         2        f   
989528        122472           99780     1         1        f   

                        name     rgb is_trans  
0           Dark Bluish Gray  6C6E68        f  
1                 Light Gray  9BA19D        f  
2       [No Color/Any Color]  05131D        f  
3                      Black  05131D        f  
4                     Orange  FE8A18        f  
...                      ...     ...      ...  
989524     Light Bluish Gray  A0A5A9        f  
989525     Light Bluish Gray  A0A5A9        f  
989526                 Black  05131D        f  
989527                 Black  05131D        f  
989528                  Blue  0055BF        f  

[989529 rows x 8 columns]

In [100]:
def get_part_colors(part_number):
    part_number = str(part_number)
    return colors_final[colors_final['part_num']==part_number][['id','name','rgb','is_trans']].drop_duplicates().sort_values(by=(['id'])).reset_index(drop=True)

In [101]:
get_part_colors('10247')

id               name     rgb is_trans
0   0              Black  05131D        f
1   4                Red  C91A09        f
2  14             Yellow  F2CD37        f
3  15              White  FFFFFF        f
4  19                Tan  E4CD9E        f
5  71  Light Bluish Gray  A0A5A9        f
6  72   Dark Bluish Gray  6C6E68        f